src/Functions/FunctionsCodingULID.cpp

```diff
@@ -0,0 +1,182 @@
+                ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
+                "Wrong number of arguments for function {}: should be 1 or 2",
+                getName());
+
+        const auto * arg_fixed_string = checkAndGetDataType<DataTypeFixedString>(arguments[0].type.get());
+        const auto * arg_string = checkAndGetDataType<DataTypeString>(arguments[0].type.get());
+
+        if (!arg_string && !(arg_fixed_string && arg_fixed_string->getN() == ULID_LENGTH))
+            throw Exception(
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                "Illegal type {} of argument of function {}. Must be String or FixedString(26).",
+                arguments[0].type->getName(),
+                getName());
+
+        String timezone;
+        if (arguments.size() == 2)
+        {
+            timezone = extractTimeZoneNameFromColumn(*arguments[1].column);
+
+            if (timezone.empty())
+                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                    "Function {} supports a 2nd argument (optional) that must be a valid time zone",
+                    getName());
+        }
+
+        return std::make_shared<DataTypeDateTime64>(DATETIME_SCALE, timezone);
+    }
+
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    {
+        auto col_res = ColumnDateTime64::create(input_rows_count, DATETIME_SCALE);
+        auto & vec_res = col_res->getData();
+
+        const ColumnPtr column = arguments[0].column;
+
+        const auto * column_fixed_string = checkAndGetColumn<ColumnFixedString>(column.get());
+        const auto * column_string = checkAndGetColumn<ColumnString>(column.get());
+
+        if (column_fixed_string)
+        {
+            if (column_fixed_string->getN() != ULID_LENGTH)
+                throw Exception(
+                    ErrorCodes::ILLEGAL_COLUMN,
+                    "Illegal column {} of argument of function {}, expected String or FixedString({})",
+                    arguments[0].name, getName(), ULID_LENGTH
+                );
+
+            const auto & vec_src = column_fixed_string->getChars();
+
+            for (size_t i = 0; i < input_rows_count; ++i)
+            {
+                DateTime64 time = decode(vec_src.data() + i * ULID_LENGTH);
+                vec_res[i] = time;
+            }
+        }
+        else if (column_string)
+        {
+            const auto & vec_src = column_string->getChars();
+            const auto & offsets_src = column_string->getOffsets();
+
+            size_t src_offset = 0;
+
+            for (size_t i = 0; i < input_rows_count; ++i)
+            {
+                DateTime64 time = 0;
+
+                size_t string_size = offsets_src[i] - src_offset;
+                if (string_size == ULID_LENGTH + 1)
+                    time = decode(vec_src.data() + src_offset);
```

Comment:
So, if string size is not `ULID_LENGTH + 1` we just return default datetime? I think it's unexpected and It's better to throw an exception in this case. If we want to get default value or Null on bad ULID, we can implement functions `ULIDStringToDateTimeOrNull/ULIDStringToDateTimeOrDefault`